Add "Using a Symfony Reproducer" post (#45)
knpuniversity/blog: use-symfony-reproducer.md

# Using a Symfony Reproducer

***TIP
This is part 2 of a 2-part _Reproducer_ series. If you haven't already, check
out [Part 1](https://symfonycasts.com/blog/symfony-reproducer) first.
***

Ok, so you've created a reproducer app for a bug you've found in Symfony. For
this article, we'll assume the reproducer you've created in part 1 is located
at `~/my-reproducer`.

## Forking the Symfony Repository

The first thing we'll do is _fork_ the [`symfony/symfony`](https://github.com/symfony/symfony)
repository on GitHub. This will give us our own copy of the Symfony repository
where we can work.

1. Visit https://github.com/symfony/symfony and click the _Fork_
   button at the top of the page.
2. Choose your personal GitHub account as the owner and keep the default name
   of `symfony`.

> 3. Leave all the other options as is and click _Create Fork_. You should now be on https://github.com/your-username/symfony.

Reviewer: I see you also say to wait in 4th step, maybe we should split this step into 2 steps, i.e. write the part

> You should now be on https://github.com/your-username/symfony.

as a separate (5th) step?

4. Wait a few seconds for the fork to be created.
5. Click the green _Code_ dropdown and under the _SSH_ tab,
   copy the URL to your clipboard.

## Cloning your Fork

Let’s get the repo down to your local workspace:

```bash
cd ~
git clone <your-fork-url-from-clipboard>
cd symfony
```

Now we’re going to add the original Symfony repo as an _upstream_ remote. Why?
So we can easily pull in the latest Symfony updates and stay fresh!

```bash
git remote add upstream git@github.com:symfony/symfony.git
```

Let’s grab the latest changes from Symfony (you’ll want to run this command
whenever Symfony gets an update, so it’s good to get used to it).

```bash
git fetch upstream
```

## Use your Reproducer to Fix the Bug

In Part 1, you picked the Symfony version for your reproducer. Now it’s time

> to match that version in our forked Symfony repo. Let’s check it out:

Reviewer: Maybe also add something like:

> If you picked 7.1 as Symfony version for your reproducer, check it out with:
>
> ```bash
> git checkout 7.1
> ```

Or maybe at least something like this:

```bash
git checkout 7.1 # or any other version you picked for your reproducer
```

```bash
git checkout 7.1
```

Boom! You’ve got the right version. Now, let’s make sure we’re up-to-date

> by merging in any new changes:

Reviewer: ```suggestion
by merging in any new changes from the upstream repository:
```

```bash
git merge upstream/7.1
```

Great! Now, we’re going to create a new branch for our bug fix:

```bash
git checkout -b my-fix
```

***TIP
The branch name (`my-fix`) isn't important, but it should be descriptive and
unique to your fork.
***

## Linking your Reproducer to your Fork

Let’s link that reproducer you created to your freshly forked Symfony repo.
Head back to your reproducer directory:

```bash
cd ~/my-reproducer
```

Symfony comes with a handy PHP script to link itself to a project. This will
symlink all your app’s `symfony/*` packages to your local fork. Super neat, right?

```bash
php ../symfony/link .
```

## Fixing the Bug

Open your reproducer in your favorite IDE. Normally, messing with files in the 

> `vendor/` directory is a no-no, but in this case, for files in `vendor/symfony/*`,

Reviewer: ```suggestion
`vendor/` directory is a no-go, but in this case, for files in `vendor/symfony/*`,
```
?

Author: I feel like *no-go* has a different meaning. Maybe this doesn't translate well to non-english speakers (or is too _childish_ 😆)?

Reviewer: Aha, that exists in my lang too... though, IMO no-no is just a childish version of no-go, so monthly both have the same meaning for me 😅  I'm fine with any option here actually, just wanted to catch some possible misprints :)

> you can! Changes you make to these files  are actually being made in your forked

Reviewer: ```suggestion
you can! Changes you make to these files are actually being made in your forked
```

> Symfony repository.

Reviewer: ```suggestion
Symfony repository thanks to the symlinks.
```

Once you’ve squashed that bug, jump back into your local Symfony fork and take a
look at your changes:

```bash
cd ~/symfony
git status
```

If everything looks good, follow the Symfony documentation on
[Creating a Pull Request](https://symfony.com/doc/current/contributing/code/pull_requests.html#step-4-submit-your-pull-request)
to submit your fix to the Symfony repository!

## Verifying a Proposed Fix

> So, you’ve found a bug, and someone from the awesome Symfony community has proposed a fix.

Reviewer: Hm, it's a bit rough transition to this section IMO. We just fixed the bug ourselves and now we're talking about someone else fixed it. Probably make it more smooth? Something like:

> But if you can't fix it yourself - don't worry! Someone from the awesome Symfony community may propose a fix. And when it will happen - you may want to help with the review and confirm with your reproducer if the fix works

Now, let’s use your reproducer to check if the fix works.

First, track down the Pull Request (PR) that has the fix. You’ll see something like this
at the top of the PR:

> {user} wants to merge X commit(s) into symfony:{version} from {user}:{branch-name}

Click `{user}:{branch-name}` to jump to the `{user}`'s fork. Like we did with
our own fork, click the green _Code_ dropdown and copy the URL to your clipboard.

We’re going to add their fork as a remote:

```bash
cd ~/symfony
git remote add {user} <fork-url-from-clipboard>
git fetch {user}
```

Now, checkout the branch from the user's fork:

```bash
git switch -c {user}-{branch-name} {user}/{branch-name}
```

***TIP

> I like to prefix the branch name with the user’s GitHub handle

Reviewer: Could we rephrase this "user's GitHub handle"? I got it intuitively, but not sure everyone will get it

# Using a Symfony Reproducer

***TIP
This is part 2 of a 2-part _Reproducer_ series. If you haven't already, check
out [Part 1](https://symfonycasts.com/blog/symfony-reproducer) first.
***

Ok, so you've created a reproducer app for a bug you've found in Symfony. For
this article, we'll assume the reproducer you've created in part 1 is located
at `~/my-reproducer`.

## Forking the Symfony Repository

> The first thing we'll do is _fork_ the [`symfony/symfony`](https://github.com/symfony/symfony)

Reviewer 2: ```suggestion
Step 1 is to _fork_ the [`symfony/symfony`](https://github.com/symfony/symfony)
```

> repository on GitHub. This will give us our own copy of the Symfony repository

Reviewer 2: ```suggestion
repository on GitHub. This will give us our own copy
```

1. Visit https://github.com/symfony/symfony and click the _Fork_
   button at the top of the page.
2. Choose your personal GitHub account as the owner and keep the default name
   of `symfony`.
3. Leave all the other options as is and click _Create Fork_.
4. You should now be on https://github.com/your-username/symfony.
5. Wait a few seconds for the fork to be created.
6. Click the green _Code_ dropdown and under the _SSH_ tab,
   copy the URL to your clipboard.

## Cloning your Fork

Let's get the repo down to your local workspace:

```bash
cd ~
git clone <your-fork-url-from-clipboard>
cd symfony
```

Now we're going to add the original Symfony repo as an _upstream_ remote. Why?
So we can easily pull in the latest Symfony updates and stay fresh!

```bash
git remote add upstream git@github.com:symfony/symfony.git
```

Let's grab the latest changes from Symfony (you'll want to run this command
whenever Symfony gets an update, so it's good to get used to it).

```bash
git fetch upstream
```

## Use your Reproducer to Fix the Bug

In Part 1, you picked the Symfony version for your reproducer. Now it's time
to match that version in our forked Symfony repo. Let's check it out:

```bash

> git checkout 7.1 # or any other version you picked for your reproducer

Reviewer 2: Does this create that branch locally? I always use with `-b`, so I may be ignorant on what's happening

Author: It does but I can show the long form:

```
git checkout -b 7.1 upstream/7.1
```

```

Boom! You've got the right version. Now, let's make sure we're up-to-date
by merging in any new changes from the _upstream_ repository:

```bash
git merge upstream/7.1
```

> Great! Now, we're going to create a new branch for our bug fix:

Reviewer 2: ```suggestion
Great! Next, create a new branch for our bug fix:
```
